src/binder/clrprivbinderassemblyloadcontext.cpp

```diff
@@ -268,9 +299,80 @@ HRESULT CLRPrivBinderAssemblyLoadContext::SetupContext(DWORD      dwAppDomainId,
     return hr;
 }
 
+#ifdef FEATURE_COLLECTIBLE_ALC
+
+/* static */
+BOOL CLRPrivBinderAssemblyLoadContext::DestroyContext(CLRPrivBinderAssemblyLoadContext *pBindContext)
+{
+    {
+        CrstHolder holder(pBindContext->m_loadersCrst);
+
+        LoaderAllocatorSet &loaderAllocators = pBindContext->m_loaderAllocators;
+
+        LoaderAllocatorSet::Iterator iter = loaderAllocators.Begin();
+        while (iter != loaderAllocators.End())
+        {
+            LoaderAllocator *pLoaderAllocator = *iter;
+
+            // This context should be holding onto a reference to the LoaderAllocator, so it
+            // should still be alive
+            _ASSERTE(pLoaderAllocator->IsAlive());
+
+            if (pLoaderAllocator->IsManagedScoutAlive())
+            {
+                // We can't destroy until there is no managed reference to any of our LoaderAllocators
+                return FALSE;
+            }
+
+            iter++;
+        }
+
+        iter = loaderAllocators.Begin();
+        while (iter != loaderAllocators.End())
+        {
+            LoaderAllocator *pLoaderAllocator = *iter;
+
+            pLoaderAllocator->ReleaseAllReferences();
+
+            // Release our reference to the LoaderAllocator so it can be deleted by the LoaderAllocator GC
+            VERIFY(pLoaderAllocator->Release());
+
+            iter++;
+        }
+    }
+
+    // TODO: Is AppDomain::GetCurrentDomain() going to work properly?
+    AppDomain *pAppDomain = AppDomain::GetCurrentDomain();
```

Comment:
The above will work only when there is a single AppDomain in the process (which happens to be the current design for CoreCLR). What you really want is to get reference to the AppDomain that is associated with the Binder.

CLRPrivBinderAssemblyLoadContext::SetupContext associates the Binder's AppContext with the AppDomain who assemblies it is managing. Use SystemDomain::GetAppDomainFromId to get the AppDomain* for AppDomainID saved in SetupContext.